build_runner_core/test/invalidation/invalidation_tester.dart

```diff
@@ -0,0 +1,306 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:build/build.dart';
+import 'package:build_test/build_test.dart';
+import 'package:built_collection/built_collection.dart';
+import 'package:crypto/crypto.dart';
+import 'package:logging/logging.dart';
+import 'package:test/test.dart';
+
+/// Test case which build steps are triggered by changes to files.
+///
+/// For concise tests, "names" are used instead of asset IDs. The name `a` maps
+/// to the path `lib/a.dart`; all names map to the package `pkg`. "Hidden" asset
+/// IDs under `.dart_tool` are mapped back to the same namespace.
+class InvalidationTester {
+  /// The source assets on disk before the first build.
+  final Set<AssetId> _sourceAssets = {};
+
+  /// The builders that will run.
+  final List<TestBuilder> _builders = [];
+
+  /// The [OutputStrategy] for generated assets.
+  ///
+  /// [OutputStrategy.inputDigest] is the default if there is no entry.
+  final Map<AssetId, OutputStrategy> _outputs = {};
+
+  /// Assets written by generators.
+  final Set<AssetId> _generatedOutputsWritten = {};
+
+  /// The [TestReaderWriter] from the most recent build.
+  TestReaderWriter? _readerWriter;
+
+  /// The build number for "printOnFailure" output.
+  int _buildNumber = 0;
+
+  /// Sets the assets that will be on disk before the first build.
+  ///
+  /// See the note on "names" in the class dartdoc.
+  void sources(Iterable<String> names) {
+    _sourceAssets.clear();
+    for (final name in names) {
+      _sourceAssets.add(name.assetId);
+    }
+  }
+
+  /// Adds a builder to the test.
+  ///
+  /// [from] and [to] are the input and output extension of the builder,
+  /// without ".dart". So `from: '', to: '.g.dart'` is a builder that takes
```

Comment:
From

```
  TestBuilder(this._tester, String from, Iterable<String> to, this.isOptional)
    : buildExtensions = {'$from.dart': to.map((to) => '$to.dart').toList()};
```

it seems that the "to" here should just be `.g`?

Reply:
Good catch, thanks. Done.